### PR TITLE
add: metrics(Grafana)

### PR DESCRIPTION
**Pull request by LiteSun**

### CODE REVIEW

File: src/pages/Metrics/Metrics.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import React, { useState } from 'react';
+import { PageHeaderWrapper } from '@ant-design/pro-layout';
+import { Empty, Button, Card } from 'antd';
+import { history } from 'umi';
+import { stringify } from 'qs';
+
+const Metrics: React.FC<{}> = () => {
```
Comment (juzhiyuan) [inline src/pages/Metrics/Metrics.tsx:7]: ,

File: src/pages/Metrics/Metrics.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  const GLOBAL_ADMIN_SETTING_GRAFANA_URL =
+    localStorage.getItem('GLOBAL_ADMIN_SETTING_GRAFANA_URL') || undefined;
```
Comment (juzhiyuan) [inline src/pages/Metrics/Metrics.tsx:9]: undefined?

File: src/pages/Metrics/Metrics.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  const [showMetricsDashboard] = useState(GLOBAL_ADMIN_SETTING_GRAFANA_URL);
```
Comment (juzhiyuan) [inline src/pages/Metrics/Metrics.tsx:10]: 1. showMetrics is fine;
2. useState(Boolean(localStorage.getItem('GLOBAL_ADMIN_SETTING_GRAFANA_URL')))

Comment (juzhiyuan) [reply src/pages/Metrics/Metrics.tsx:10]: or `const grafanaURL = xxx`

File: src/pages/Setting/typingd.d.ts [hunk 4]
```diff
@@ -5,4 +5,8 @@ declare namespace Setting {
     path: string;
     key: string;
   }
+
+  interface DashboardConfig extends AdminAPI {
```
Comment (juzhiyuan) [inline src/pages/Setting/typingd.d.ts:9]: ,

File: src/pages/Setting/service.ts [hunk 5]
```diff
@@ -1,8 +1,9 @@
-export const getAdminAPIConfig = (): Setting.AdminAPI => {
+export const getAdminAPIConfig = (): Setting.DashboardConfig => {
```
Comment (juzhiyuan) [inline src/pages/Setting/service.ts:1]: A & B

File: src/pages/Setting/service.ts [hunk 6]
```diff
@@ -1,8 +1,9 @@
...
   return {
     schema: localStorage.getItem('GLOBAL_ADMIN_API_SCHEMA') || 'http',
     host: localStorage.getItem('GLOBAL_ADMIN_API_HOST') || '127.0.0.1:9080',
     path: localStorage.getItem('GLOBAL_ADMIN_API_PATH') || '/apisix/admin',
     key: localStorage.getItem('GLOBAL_ADMIN_API_KEY') || '',
+    grafanaUrl: localStorage.getItem('GLOBAL_ADMIN_SETTING_GRAFANA_URL') || '',
```
Comment (juzhiyuan) [inline src/pages/Setting/service.ts:7]: grafanaURL

File: src/pages/Metrics/Metrics.tsx (hunk 2 above)
Comment (juzhiyuan) [reply src/pages/Metrics/Metrics.tsx:9]: from service?

File: src/pages/Metrics/Metrics.tsx [hunk 7]
```diff
@@ -1,17 +1,24 @@
...
 import { Empty, Button, Card } from 'antd';
 import { history } from 'umi';
 import { stringify } from 'qs';
+import { getGrafanaConfig } from './service';
 
 const Metrics: React.FC<{}> = () => {
-  const GLOBAL_ADMIN_SETTING_GRAFANA_URL =
-    localStorage.getItem('GLOBAL_ADMIN_SETTING_GRAFANA_URL') || undefined;
-  const [showMetricsDashboard] = useState(GLOBAL_ADMIN_SETTING_GRAFANA_URL);
+  const [grafanaURL, setGrafanaURL] = useState<string>('');
```
Comment (juzhiyuan) [inline src/pages/Metrics/Metrics.tsx:9]: const [] = useState<string | undefined>()

File: src/pages/Metrics/Metrics.tsx [hunk 8]
```diff
@@ -1,17 +1,24 @@
...
+  const [showMetrics, setShowMetrics] = useState(Boolean(grafanaURL));
+
+  useEffect(() => {
+    const url = getGrafanaConfig();
```
Comment (juzhiyuan) [inline src/pages/Metrics/Metrics.tsx:13]: getXXConfig should return an Object but not a fixed value...

File: src/pages/Metrics/Metrics.tsx [hunk 9]
```diff
@@ -1,17 +1,24 @@
...
 import { history } from 'umi';
 import { stringify } from 'qs';
+import { getGrafanaConfig } from './service';
 
 const Metrics: React.FC<{}> = () => {
-  const GLOBAL_ADMIN_SETTING_GRAFANA_URL =
-    localStorage.getItem('GLOBAL_ADMIN_SETTING_GRAFANA_URL') || undefined;
-  const [showMetricsDashboard] = useState(GLOBAL_ADMIN_SETTING_GRAFANA_URL);
+  const [grafanaURL, setGrafanaURL] = useState<string>('');
+  const [showMetrics, setShowMetrics] = useState(Boolean(grafanaURL));
```
Comment (juzhiyuan) [inline src/pages/Metrics/Metrics.tsx:10]: always false

File: src/pages/Metrics/service.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+export const getGrafanaConfig = (): string => {
+  return localStorage.getItem('GLOBAL_ADMIN_SETTING_GRAFANA_URL') || '';
```
Comment (juzhiyuan) [inline src/pages/Metrics/service.ts:2]: ,

File: src/pages/Setting/Setting.tsx [hunk 11]
```diff
@@ -14,15 +14,21 @@ const Settings: React.FC<{}> = () => {
   const { formatMessage } = useIntl();
 
   useEffect(() => {
-    form.setFieldsValue(getAdminAPIConfig());
+    form.setFieldsValue({ ...getAdminAPIConfig(), ...getGrafanaConfig() });
```
Comment (juzhiyuan) [inline src/pages/Setting/Setting.tsx:17]: ,

File: src/pages/Setting/typingd.d.ts [hunk 12]
```diff
@@ -6,7 +6,7 @@ declare namespace Setting {
     key: string;
   }
 
-  interface DashboardConfig extends AdminAPI {
-    grafanaUrl: string;
+  interface GrafanaURL {
```
Comment (juzhiyuan) [inline src/pages/Setting/typingd.d.ts:9]: too wired...

File: src/pages/Metrics/Metrics.tsx (hunk 1 above)
Comment (juzhiyuan) [reply src/pages/Metrics/Metrics.tsx:8]: empty typing should be remove

File: src/pages/Setting/Setting.tsx (hunk 11 above)
Comment (LiteSun) [reply src/pages/Setting/Setting.tsx:17]: ?

File: src/pages/Setting/typingd.d.ts (hunk 12 above)
Comment (LiteSun) [reply src/pages/Setting/typingd.d.ts:9]: ?

File: src/pages/Metrics/Metrics.tsx (hunk 1 above)
Comment (LiteSun) [reply src/pages/Metrics/Metrics.tsx:8]: should I remove `React.FC<{}>` in `pages/404.tsx` `pages/Setting/Setting.tsx` file?

Comment (juzhiyuan) [reply src/pages/Metrics/Metrics.tsx]: yep but not in this PR, those files were generated by Ant Design Pro.

File: src/pages/Setting/typingd.d.ts (hunk 12 above)
Comment (juzhiyuan) [reply src/pages/Setting/typingd.d.ts]: Why the typing GrafanaURL includes a grafanaURL property??

File: src/pages/Setting/typingd.d.ts [hunk 13]
```diff
@@ -6,7 +6,7 @@ declare namespace Setting {
     key: string;
   }
 
-  interface GrafanaURL {
+  interface GrafanaConfig {
     grafanaURL: string;
```
Comment (juzhiyuan) [inline src/pages/Setting/typingd.d.ts:10]: that's it..

**Pull request merged**